### PR TITLE
BigQuery deferrable operators

### PR DESCRIPTION
**Pull request by wojsamjan**
Add BigQuery deferrable operators

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: tests/providers/google/cloud/hooks/test_bigquery.py [hunk 1]
```diff
@@ -2053,3 +2055,195 @@ def test_create_external_table_description(self, mock_create):
...
         _, kwargs = mock_create.call_args
         assert kwargs['table_resource']['description'] is description
+
+
+@pytest.mark.asyncio
+@mock.patch("airflow.providers.google.cloud.hooks.bigquery.ClientSession")
+async def test_get_job_instance(mock_session):
+    hook = BigQueryHookAsync()
+    result = await hook.get_job_instance(project_id=PROJECT_ID, job_id=JOB_ID, session=mock_session)
+    assert isinstance(result, Job)
```
Comment (kaxil) [inline tests/providers/google/cloud/hooks/test_bigquery.py:2065]: https://github.com/astronomer/astronomer-providers/blob/main/tests/google/cloud/hooks/test_bigquery.py#L74-L79

Tests are also copied from astronomer providers.

**Review by kaxil — CHANGES_REQUESTED** [review]
While I appreciate your contribution, the code is copy/paste of Bigquery Operators in astronomer-providers repo.

I don't appreciate the fact that you did not credit the source of the code in the PR description, this is against the spirit of open-source. We are planning to donate the code in apache/airflow repo anyways after 6 months of the code being released -- we will raise separate PRs for contributing it in next few days.

File: airflow/providers/google/common/hooks/base_google.py [hunk 2]
```diff
@@ -602,3 +603,26 @@ def test_connection(self):
...
+        Sync version of the Google Cloud Hooks makes blocking calls in ``__init__`` so we don't inherit
+        from it.
+        """
+        if not self._sync_hook:
+            self._sync_hook = await sync_to_async(self.sync_hook_class)(**self._hook_kwargs)
+        return self._sync_hook
+
+    async def service_file_as_context(self) -> Any:
+        sync_hook = await self.get_sync_hook()
+        return await sync_to_async(sync_hook.provide_gcp_credential_file_as_context)()
```
Comment (kaxil) [inline airflow/providers/google/common/hooks/base_google.py:628]: Same code as 
https://github.com/astronomer/astronomer-providers/blob/main/astronomer/providers/google/common/hooks/base_google.py#L7-L27

File: airflow/providers/google/cloud/hooks/bigquery.py [hunk 3]
```diff
@@ -29,8 +29,10 @@
 import warnings
 from copy import deepcopy
 from datetime import datetime, timedelta
-from typing import Any, Dict, Iterable, List, Mapping, NoReturn, Optional, Sequence, Tuple, Type, Union
+from typing import Any, Dict, Iterable, List, Mapping, NoReturn, Optional, Sequence, Tuple, Type, Union, cast
 
+from aiohttp import ClientSession as ClientSession
```
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/bigquery.py:34]:
```suggestion
from aiohttp import ClientSession
```

File: airflow/providers/google/cloud/hooks/bigquery.py [hunk 4]
```diff
@@ -2821,6 +2824,221 @@ def _get_query_result(self) -> Dict:
...
+    """BigQueryHookAsync inherits from GoogleBaseHookAsync class. Interacts with Google BigQuery"""
+
+    sync_hook_class = BigQueryHook
+
+    async def get_job_instance(
+        self, project_id: Optional[str], job_id: Optional[str], session: ClientSession
+    ) -> Job:
+        """Get the specified job resource by job ID and project ID."""
+        with await self.service_file_as_context() as f:
+            return Job(job_id=job_id, project=project_id, service_file=f, session=cast(Session, session))
```
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/bigquery.py:2837]: This `cast` doesn’t feel right… do you have a reference for this?

File: airflow/providers/google/cloud/hooks/bigquery.py [hunk 5]
```diff
@@ -2821,6 +2824,221 @@ def _get_query_result(self) -> Dict:
...
+        self,
+        sql: str,
+        pass_value: Any,
+        records: List[Any],
+        tolerance: Optional[float] = None,
+    ) -> None:
+        """
+        Match a single query resulting row and tolerance with pass_value
+
+        :return: Raises AirflowException if there is no match.
```
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/bigquery.py:2896]:
```suggestion
        :raises: AirflowException if there is no match.
```

**Comment by kaxil** [comment]
Closing this in favor of https://github.com/apache/airflow/pull/26156

**Pull request closed without merging**